src/cdef_apply.rs

```diff
@@ -1,270 +1,226 @@
 use crate::include::common::bitdepth::BitDepth;
+use crate::include::common::bitdepth::BPC;
 use crate::include::common::intops::ulog2;
 use crate::include::dav1d::headers::Rav1dPixelLayout;
 use crate::src::align::Align16;
 use crate::src::cdef::CdefEdgeFlags;
-use crate::src::cdef::CDEF_HAVE_BOTTOM;
-use crate::src::cdef::CDEF_HAVE_LEFT;
-use crate::src::cdef::CDEF_HAVE_RIGHT;
-use crate::src::cdef::CDEF_HAVE_TOP;
 use crate::src::internal::Rav1dContext;
 use crate::src::internal::Rav1dDSPContext;
 use crate::src::internal::Rav1dFrameContext;
 use crate::src::internal::Rav1dTaskContext;
 use crate::src::lf_mask::Av1Filter;
+use bitflags::bitflags;
 use libc::ptrdiff_t;
 use std::cmp;
 use std::ffi::c_int;
 use std::ffi::c_uint;
 use std::slice;
 
-pub type Backup2x8Flags = c_uint;
-pub const BACKUP_2X8_UV: Backup2x8Flags = 2;
-pub const BACKUP_2X8_Y: Backup2x8Flags = 1;
+bitflags! {
+    #[derive(Clone, Copy)]
+    struct Backup2x8Flags: u8 {
+        const Y = 1 << 0;
+        const UV = 1 << 1;
+    }
+}
+
+impl Backup2x8Flags {
+    pub const fn select(&self, select: bool) -> Self {
+        if select {
+            *self
+        } else {
+            Self::empty()
+        }
+    }
+}
 
 unsafe fn backup2lines<BD: BitDepth>(
-    dst: *const *mut BD::Pixel,
-    src: *const *mut BD::Pixel,
-    stride: *const ptrdiff_t,
+    dst: &[*mut BD::Pixel],
+    src: &[*mut BD::Pixel; 3],
+    stride: &[ptrdiff_t; 2],
     layout: Rav1dPixelLayout,
 ) {
-    let y_stride: ptrdiff_t = BD::pxstride(*stride.offset(0) as usize) as isize;
+    let y_stride: ptrdiff_t = BD::pxstride(stride[0] as usize) as isize;
     if y_stride < 0 {
+        let len = (-2 * y_stride) as usize;
         BD::pixel_copy(
-            slice::from_raw_parts_mut(
-                *dst.offset(y_stride as isize) as *mut BD::Pixel,
-                (-2 * y_stride) as usize,
-            ),
-            slice::from_raw_parts(
-                *src.offset((7 as isize * y_stride) as isize) as *const BD::Pixel,
-                (-2 * y_stride) as usize,
-            ),
-            (-2 * y_stride) as usize,
+            slice::from_raw_parts_mut(dst[0].offset(y_stride), len),
+            slice::from_raw_parts(src[0].offset(7 * y_stride), len),
```

Comment:
I think these were mistakenly mistranslated before, as `src` and `dst` are offset first and then dereferened, vs. dereferenced/`[0]`ed first, and then offset.  Your way looks correct so all is good here.  I'll add to the PR description that this bug is fixed as well.